### PR TITLE
Extract Explain, Analyze, Extension in LogicalPlan as independent struct

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1307

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs [hunk 1]
```diff
@@ -973,24 +973,24 @@ impl TryInto<protobuf::LogicalPlanNode> for &LogicalPlan {
                     )),
                 })
             }
-            LogicalPlan::Analyze { verbose, input, .. } => {
-                let input: protobuf::LogicalPlanNode = input.as_ref().try_into()?;
+            LogicalPlan::Analyze(a) => {
```
Comment (liukun4515) [inline ballista/rust/core/src/serde/logical_plan/to_proto.rs:976]: It's better to import the Analyze struct  and use this code style 
```
LogicalPlan::Analyze(Aanlyze{
 verbose, input,....
})
```

File: datafusion/src/execution/context.rs [hunk 2]
```diff
@@ -650,28 +651,23 @@ impl ExecutionContext {
...
     pub fn optimize(&self, plan: &LogicalPlan) -> Result<LogicalPlan> {
-        if let LogicalPlan::Explain {
-            verbose,
-            plan,
-            stringified_plans,
-            schema,
-        } = plan
-        {
-            let mut stringified_plans = stringified_plans.clone();
+        if let LogicalPlan::Explain(e) = plan {
```
Comment (liukun4515) [inline datafusion/src/execution/context.rs:654]: same comments like above.

File: datafusion/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -696,21 +698,21 @@ impl LogicalPlanBuilder {
         let schema = schema.to_dfschema_ref()?;
 
         if analyze {
-            Ok(Self::from(LogicalPlan::Analyze {
+            Ok(Self::from(LogicalPlan::Analyze(AnalyzePlan {
```
Comment (liukun4515) [inline datafusion/src/logical_plan/builder.rs:701]: I think this code style is better than above style.
This is same with this pull request #1290

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs (hunk 1 above)
Comment (xudong963) [reply ballista/rust/core/src/serde/logical_plan/to_proto.rs:976]: In fact, I don't have a strong desire which is better. When the struct has many variables, I'd like to use `StructVar(s) => {s.v1, s.v2 ...}`

Comment (alamb) [reply ballista/rust/core/src/serde/logical_plan/to_proto.rs:976]: I agree either style seems fine to me

**Review by alamb — APPROVED** [review]
LGTM -- thanks @xudong963 and @liukun4515  

I also merged `apache/master` into this branch locally to test for logical conflicts and things look good 👍

**Pull request merged**